ably/realtime_conn.go

```diff
@@ -0,0 +1,367 @@
+	defer c.state.Unlock()
+	return c.id
+}
+
+// Key gives unique key string obtained from Ably upon successful connection.
+// The key may change due to reconnection and recovery; on every received
+// StatConnConnected event previously obtained Key is no longer valid.
+func (c *Conn) Key() string {
+	c.state.Lock()
+	defer c.state.Unlock()
+	return c.key
+}
+
+// Ping issues a ping request against configured endpoint and returns TTR times
+// for ping request and pong response.
+//
+// Ping returns non-nil error without any attemp of communication with Ably
+// if the connection state is StateConnClosed or StateConnFailed.
+func (c *Conn) Ping() (ping, pong time.Duration, err error) {
+	return 0, 0, errors.New("TODO")
+}
+
+// Reason gives last known error that caused connection transit to
+// StateConnFailed state.
+func (c *Conn) Reason() error {
+	c.state.Lock()
+	err := c.state.err
+	c.state.Unlock()
+	return err
+}
+
+// Serial gives serial number of a message received most recently. Last known
+// serial number is used when recovering connection state.
+func (c *Conn) Serial() int64 {
+	c.state.Lock()
+	serial := c.serial
+	c.state.Unlock()
+	return serial
+}
+
+// State returns current state of the connection.
+func (c *Conn) State() int {
+	c.state.Lock()
+	state := c.state.current
+	c.state.Unlock()
+	return state
+}
+
+// On relays request connection states to the given channel; on state transition
+// connection will not block sending to c - the caller must ensure the incoming
+// values are read at proper pace or the c is sufficiently buffered.
+//
+// If no states are given, c is registered for all of them.
+// If c is nil, the method panics.
+// If c is alreadt registered, its state set is expanded.
+func (c *Conn) On(ch chan<- State, states ...int) {
+	c.state.on(ch, states...)
+}
+
+// Off removes c from listetning on the given connection state transitions.
+//
+// If no states are given, c is removed for all of the connection's states.
+// If c is nil, the method panics.
+// If c was not registered or is already removed, the method is a nop.
+func (c *Conn) Off(ch chan<- State, states ...int) {
+	c.state.off(ch, states...)
+}
+
+var errQueueing = &Error{Code: 40000, Err: errors.New("unable to send messages in current state with disabled queueing")}
+
+func (c *Conn) send(msg *proto.ProtocolMessage, listen chan<- error) error {
+	const maxint64 = 1<<63 - 1
+	c.state.Lock()
+	switch c.state.current {
+	case StateConnInitialized, StateConnConnecting, StateConnDisconnected:
+		c.state.Unlock()
+		if c.opts.NoQueueing {
+			return errQueueing
+		}
+		c.queue.Enqueue(msg, listen)
+		return nil
+	case StateConnConnected:
+	default:
+		c.state.Unlock()
+		return &Error{Code: 80000}
+	}
+	msg.MsgSerial = c.msgSerial
+	c.msgSerial = (c.msgSerial + 1) % maxint64
+	if listen != nil {
+		c.pending.Enqueue(msg.MsgSerial, listen)
+	}
+	c.state.Unlock()
+	return c.conn.Send(msg)
+}
+
+func (c *Conn) isActive() bool {
+	c.state.Lock()
```

Comment:
Firstly RWMutex is slower than Mutex, secondly there's no much to be gain from RLocks on those simple getters - majority of the lock usage is write-access, RWMutex is usefull for single-writer, multiple-readers scenarios.

Before we start to micro-optimize I'd prefer to write benchmarks first so I can be sure about the outcome; probably not going to happen before the functionality is ready first ;)